pufferlib/ocean/g2048/2048.h

```diff
@@ -0,0 +1,324 @@
+#include <stdlib.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <time.h>
+#include "raylib.h"
+
+#define SIZE 4
+#define EMPTY 0
+#define UP 1
+#define DOWN 2
+#define LEFT 3
+#define RIGHT 4
+
+typedef struct {
+    float perf;
+    float score;
+    float episode_return;
+    float episode_length;
+    float n;
+} Log;
+
+// Required struct for env_binding.h compatibility
+typedef struct {
+    Log log;                        // Required
+    float* observations;            // Required (flattened 256 floats: 16 tiles * 16 one-hot)
+    int* actions;                   // Required
+    float* rewards;                 // Required
+    unsigned char* terminals;       // Required
+    int score;
+    int tick;
+    float grid[SIZE][SIZE];         // Store tile values directly as floats
+    float episode_reward;           // Accumulate episode reward
+} Game;
+
+// --- Logging ---
+void add_log(Game* game);
+
+// --- Required functions for env_binding.h ---
+void c_reset(Game* env);
+void c_step(Game* env);
+void c_render(Game* env);
+void c_close(Game* env);
+
+// --- Helper: one-hot encode a tile value into a 16-length vector
+static void one_hot_tile(float* out, int value) {

```

Comment:
Do this in the neural net, not in the env. Reason is you blow up the size of observations and increase GPU bandwidth. Best would uint8 obs where it's log2(value), then one-hot x 16 in network

Reply:
Also be sure to change the single_observation_space dtype to match in the .py